File: exercises/cpp/lab04/priority_scheduling.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

struct Process {
    int id, burst, priority, waiting, turnaround;
};

int main() {
    int n;
    cout << "=== Priority Scheduling ===\n";
    cout << "Enter total number of processes: ";
    cin >> n;

    vector<Process> procs(n);
    cout << "\nEnter Burst Time and Priority for each process:\n";
    for (int i = 0; i < n; i++) {
        procs[i].id = i + 1;
        cout << "P[" << i + 1 << "] Burst Time: ";
        cin >> procs[i].burst;
        cout << "P[" << i + 1 << "] Priority:   ";
        cin >> procs[i].priority;
    }

    // Sort by priority (lower number = higher priority)
    sort(procs.begin(), procs.end(), [](const Process& a, const Process& b) {
        return a.priority < b.priority;
    });

    // Calculate waiting times
    procs[0].waiting = 0;
    for (int i = 1; i < n; i++)
        procs[i].waiting = procs[i-1].waiting + procs[i-1].burst;

    int total_wt = 0, total_tat = 0;
    cout << "\nProcess\tBurst Time\tPriority\tWaiting Time\tTurnaround Time\n";
    for (int i = 0; i < n; i++) {
        procs[i].turnaround = procs[i].burst + procs[i].waiting;
        total_wt  += procs[i].waiting;
        total_tat += procs[i].turnaround;
        cout << "P[" << procs[i].id << "]\t" << procs[i].burst << "\t\t"
             << procs[i].priority << "\t\t" << procs[i].waiting
             << "\t\t" << procs[i].turnaround << "\n";
    }

    cout << "\nAverage Waiting Time    = " << (float)total_wt  / n << "\n";
    cout << "Average Turnaround Time = " << (float)total_tat / n << "\n";
    return 0;
}

File: exercises/python/lab04/priority_scheduling.py
"""Lab 04B - Priority Scheduling"""

def priority_scheduling():
    print("=== Priority Scheduling ===\n")
    n = int(input("Enter total number of processes: "))

    procs = []
    print("\nEnter Burst Time and Priority for each process:")
    for i in range(n):
        burst    = int(input(f"P[{i+1}] Burst Time: "))
        priority = int(input(f"P[{i+1}] Priority:   "))
        procs.append({"id": i+1, "burst": burst, "priority": priority})

    # Sort by priority (lower = higher priority)
    procs.sort(key=lambda p: p["priority"])

    # Compute waiting times
    procs[0]["waiting"] = 0
    for i in range(1, n):
        procs[i]["waiting"] = procs[i-1]["waiting"] + procs[i-1]["burst"]

    print(f"\n{'Process':<10}{'Burst':^10}{'Priority':^12}{'Waiting':^12}{'Turnaround'}")
    total_wt = total_tat = 0
    for p in procs:
        p["tat"] = p["burst"] + p["waiting"]
        total_wt  += p["waiting"]
        total_tat += p["tat"]
        print(f"P[{p['id']}]{'':<6}{p['burst']:^10}{p['priority']:^12}{p['waiting']:^12}{p['tat']}")

    print(f"\nAverage Waiting Time    = {total_wt/n:.2f}")
    print(f"Average Turnaround Time = {total_tat/n:.2f}")


if __name__ == "__main__":
    priority_scheduling()
